### PR TITLE
feat(gear-programs): Wrapped Vara (tokenizer) program extends VftService

### PR DESCRIPTION
**Pull request by vobradovich**

### CODE REVIEW

File: gear-programs/vara-tokenizer/app/src/tokenizer_service.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        }
+
+        let to = msg::source();
+        if let Err(err) = vft_funcs::mint(to, value.into()) {
+            // TODO reply with value `program::send_reply_with_value` when `sails` allows it
+            msg::send_bytes_with_gas(to, vec![], 0, value)
+                .expect("Failed to send message with value");
+            Err(err)
+        } else {
+            // drop send event result
```
Comment (mertwole) [inline gear-programs/vara-tokenizer/app/src/tokenizer_service.rs:45]:
```suggestion
```
Or describe **why** we should do it

File: gear-programs/vara-tokenizer/app/src/tokenizer_service.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    pub async fn burn_and_return_value(&mut self, value: u128) -> Result<u128> {
+        if value == 0 {
+            return Ok(value);
+        }
+
+        let from = msg::source();
+        vft_funcs::burn(from, value.into())?;
+
+        // drop send event result
```
Comment (mertwole) [inline gear-programs/vara-tokenizer/app/src/tokenizer_service.rs:59]: Same as for `mint_from_value`

File: gear-programs/vara-tokenizer/app/src/tokenizer_service.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        let value = msg::value();
+        if value == 0 {
+            return Ok(value);
+        }
+
+        let to = msg::source();
+        if let Err(err) = vft_funcs::mint(to, value.into()) {
+            // TODO reply with value `program::send_reply_with_value` when `sails` allows it
+            msg::send_bytes_with_gas(to, vec![], 0, value)
+                .expect("Failed to send message with value");
```
Comment (mertwole) [inline gear-programs/vara-tokenizer/app/src/tokenizer_service.rs:42]:
```suggestion
                .expect("Failed to send value to user");
```

File: gear-programs/vft-client/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+[package]
+name = "vft-client"
+version.workspace = true
+edition.workspace = true
+
+[dependencies]
+mockall = { version = "0.12", optional = true }
```
Comment (mertwole) [inline gear-programs/vft-client/Cargo.toml:7]: please move it to workspace deps

Comment (mertwole) [reply gear-programs/vft-client/Cargo.toml:7]: The same for all the `Cargo.toml`s

File: gear-programs/vara-tokenizer/tests/gclient.rs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    (api, remoting, code_id)
+}
+
+#[tokio::test]
+#[ignore = "requires run gear node on GEAR_PATH"]
+async fn factory_works() {
+    // arrange
+    let (api, remoting, program_code_id) = init_remoting().await;
+    let _admin_id =
+        ActorId::try_from(api.account_id().encode().as_ref()).expect("failed to create actor id");
```
Comment (mertwole) [inline gear-programs/vara-tokenizer/tests/gclient.rs:23]:
```suggestion
```

File: gear-programs/vara-tokenizer/tests/gclient.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+    let program_id = program_factory
+        .new("Name".into(), "Symbol".into(), 10u8)
+        .send_recv(program_code_id, b"salt")
+        .await
+        .unwrap();
+
+    let _initial_balance = api
+        .free_balance(admin_id)
+        .await
+        .expect("Failed to get free balance");
```
Comment (mertwole) [inline gear-programs/vara-tokenizer/tests/gclient.rs:114]:
```suggestion
```

File: gear-programs/vara-tokenizer/tests/gtest.rs [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+    let client_balance = client
+        .balance_of(ADMIN_ID.into())
+        .recv(program_id)
+        .await
+        .unwrap();
+
+    let balance = remoting.system().balance_of(ADMIN_ID);
+    let program_balance = remoting.system().balance_of(program_id);
+    // TODO update test after next `gtest` release
```
Comment (mertwole) [inline gear-programs/vara-tokenizer/tests/gtest.rs:124]: Please describe what changes you expect in new release or link an issue, if possible

File: gear-programs/vara-tokenizer/app/src/tokenizer_service.rs [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    pub async fn mint_from_value(&mut self) -> Result<u128> {
+        let value = msg::value();
+        if value == 0 {
+            return Ok(value);
+        }
+
+        let to = msg::source();
+        if let Err(err) = vft_funcs::mint(to, value.into()) {
+            // TODO reply with value `program::send_reply_with_value` when `sails` allows it
```
[the remark this answers is not in the document]
Comment (vobradovich) [reply gear-programs/vara-tokenizer/app/src/tokenizer_service.rs:40]: https://github.com/gear-tech/sails/issues/475

[the remark this answers is not in the document]
Comment (mertwole) [reply gear-programs/vara-tokenizer/app/src/tokenizer_service.rs:40]: I mean in the code

File: gear-programs/vara-tokenizer/Cargo.toml [hunk 9]
```diff
@@ -0,0 +1,25 @@
+[package]
+name = "vara-tokenizer"
+version.workspace = true
+edition.workspace = true
+
+[dependencies]
+vara-tokenizer-app = { path = "app" }
```
Comment (gshep) [inline gear-programs/vara-tokenizer/Cargo.toml:7]: I would place that and all its deps (client, etc) in the workspace's Cargo.toml and use like `.workspace = true`

File: gear-programs/vara-tokenizer/app/src/admin_service.rs [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    gstd::msg,
+    prelude::{collections::HashSet, *},
+};
+
+#[derive(Debug)]
+pub(crate) struct AdminConfig {
+    pub admins: HashSet<ActorId>,
+}
+
+static_storage!(AdminConfig);
```
Comment (gshep) [inline gear-programs/vara-tokenizer/app/src/admin_service.rs:11]: It is better to keep it in the program's state from my point of view

File: gear-programs/vara-tokenizer/tests/gtest.rs [hunk 11]
```diff
@@ -0,0 +1,177 @@
+use sails_rs::{calls::*, gtest::calls::*, prelude::*};
```
Comment (gshep) [inline gear-programs/vara-tokenizer/tests/gtest.rs:1]: do we really need to have both tests for gclient and gtest? To tell the truth I personally prefer gclient only

**Review by gshep — APPROVED** [review]
:+1:

File: gear-programs/vara-tokenizer/Cargo.toml (hunk 9 above)
Comment (vobradovich) [reply gear-programs/vara-tokenizer/Cargo.toml:7]: It is from `sails` template. Main crate referenced in the workspace's Cargo.toml. The `-app' crate is intended for internal use only.

File: gear-programs/vara-tokenizer/tests/gtest.rs (hunk 11 above)
Comment (vobradovich) [reply gear-programs/vara-tokenizer/tests/gtest.rs:1]: But the `gclient` tests require a running node and a lot of time to run.
So they marked as `ignored` by default.
I think we should have both.

File: gear-programs/vara-tokenizer/Cargo.toml (hunk 9 above)
Comment (vobradovich) [reply gear-programs/vara-tokenizer/Cargo.toml:7]: But it makes sense if you're referring to a service from another.

File: gear-programs/vara-tokenizer/app/src/admin_service.rs (hunk 10 above)
Comment (vobradovich) [reply gear-programs/vara-tokenizer/app/src/admin_service.rs:11]: Methods `storage` and `storage_mut` are generated by macro, which may conflict with other storages.
It is better to use in the service module.

**Review by mertwole — COMMENTED** [review]
Waiting for
- repository for `vft` service to depend on instead of copying code here
- https://github.com/gear-tech/sails/issues/475
as discussed in DM

File: gear-programs/vara-tokenizer/Cargo.toml (hunk 9 above)
Comment (mertwole) [reply gear-programs/vara-tokenizer/Cargo.toml]: @gshep @vobradovich What's with this review?

Comment (vobradovich) [reply gear-programs/vara-tokenizer/Cargo.toml]: all crates added to workspace's Cargo.toml

File: gear-programs/vft-service/Cargo.toml [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+version.workspace = true
+edition.workspace = true
+
+[dependencies]
+gstd.workspace = true
+log = "*"
+sails-rs.workspace = true
+
+[dev-dependencies]
+env_logger = "*"
```
Comment (mertwole) [inline gear-programs/vft-service/Cargo.toml:12]: Workspace deps

**Review by mertwole — APPROVED** [review]

**Pull request merged**